### DISCUSSION TITLE
Extend Validation to FK References

### DISCUSSION THREAD

**Discussion by rogersanick** [directus/directus#11166] [Feature Requests]
Hi there! 

My team would very much benefit from functionality enabling validation on referenced entities. 

An simple example would be:

- We have an entity called `Classroom`
- `Classroom` has a `VARCHAR` field which is configured to be a dropdown with either `FULL` or `NOT_FULL`
- Classrooms have five columns each containing a FK referencing a record in the students table 
- The Students table has a BOOL column indicating whether or not a given student's tuition is paid

A constraint based on referenced FKs might look like:
- A classroom can only be marked as full when all five columns reference a student and each student has paid their tuition

We use this structure in our current CMS (Sanity) to validate the status of many different hierarchically related entities. It would be incredibly useful to have that available in Directus.

**Comment by rijkvanzanten** [comment]
How / where would you set that up ideally? Do you have an example of what that could look like in Directus' filter structure?

Something like

```json
{
  "students": {
    "_each": {
      "tuition_paid": {
        "_eq": true
      }
    }
  }
}
```

??

**Reply by rogersanick** [reply]
![Screen Shot 2022-01-24 at 3 45 43 PM](https://user-images.githubusercontent.com/16504493/150861946-a63c91e5-87eb-4a80-821c-d3437c2d4297.png)

Hey @rijkvanzanten - thanks for the response. I haven't looked into the way Directus represents filters in JSON but that's super helpful.

Ideally we could keep the composeable structure with minimal changes (see attached fields)

Maybe something like: 

```
{
  all: {
  "student_ref_1": {
    "field_ref": {
      "tuition_paid": {
        "_eq": true
       }
     }
   }
 }
}
```
Where student is a FK reference.

**Reply by rijkvanzanten** [reply]
Yeah! Also, would you be able to share some insight in how you've configured it in Sanity? I'm not too familiar with their permissions management 🙂

**Reply by rogersanick** [reply]
Absolutely! 

In Sanity, the schema (and associated validation) are all explicitly configured in Typescript. The underlying document structure is based in a no-sql design. They also expose a generic client which lets you run arbitrary queries on existing data to determine if an input is valid.

I actually tried doing the same in Directus in a custom component. It worked briefly by accessing the User Store, retrieving an API key and then using that to configure the client, but I think this [release](https://github.com/directus/directus/releases/tag/v9.4.2) made that workaround impossible.

I'm going to add that as a feature request as well! I think it would be great for other use cases.

**Reply by rijkvanzanten** [reply]
Ahhh, so you basically built the validation rules for this setup yourself? 🤔

**Reply by rogersanick** [reply]
Exactly - pros were obviously that it was highly configurable, but the big cons were the fact that the dev team had to handle every validation change. 

They had the concept of both `field` and `document` level validation which were both super helpful.

**Reply by rijkvanzanten** [reply]
Interesting! Thanks for that insight.

As for field vs document level validation, we currently only do document level validation on the role level. We're currently doing the R&D on adding field-level validation (regardless of role) for more "standardized input" level validation (like making sure a field is a valid email address, etc) 🙂

**Reply by rogersanick** [reply]
No problem at all! That's super interesting, what do you mean `document level validation on the role level.` are there docs on this? This might solve a lot of my issues here. 

Also - I'm happy to take a crack at the above feature request as a PR! I would just need some review and potentially to ping you with any specific Q's that come up :)

**Reply by rijkvanzanten** [reply]
> That's super interesting, what do you mean document level validation on the role level. are there docs on this? This might solve a lot of my issues here.

The current way validation works is that you can configure a ruleset on the role level that will validate against the whole object. That will allow you to set things like:

```
_and 
  title
    _nnull true
  publish_date
    _gte $NOW
```

These rules are checked against the whole item payload, rather than individual fields (that allows for things like and/or across fields). We're currently working on an additional feature that will allow you to configure field-level validation rules on the fields themselves, so they act like a global field constraint rather than a role level validation. The idea is that the field-level validation can be used for "data-type" validation (like ensuring a value is always one of "draft", "published"), and role validation can be used to make sure that "interns" can't create "articles" that are "published".

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
